### ISSUE TITLE
Simple unite() operation fails

### ISSUE THREAD

**Issue by iconexperience** [paperjs/paper.js#1054]
Here is a simple case where the `unite()` operation still results in an open path.

```javascript
var p1 = new Path({segments:[
    [310, 170], 
    [200, 170], 
    [200, 350], 
    [190, 350], 
    [190, 160], 
    [320, 160],
    [320, 350], 
    [310, 350]
    ], closed:true, strokeColor: 'crimson', strokeScaling: false});
var p2 = new Path({segments:[
    [195, 165], 
    [315, 165], 
    [315, 350], 
    [195, 350]
    ], closed:true, strokeColor: 'teal', strokeScaling: false});
var result = p1.unite(p2);
```

![image](https://cloud.githubusercontent.com/assets/2122773/15501267/e154c53a-21ac-11e6-9664-3035d73d3237.png)

<a href="http://sketch.paperjs.org/#S/jZHRasMgFIZfRbxpB6HEhGzU0qu9wCCXSS+cOW0lJxrUtIzSd5+mC2QjlHl3vl853483qkUHlNOyBS/PNKHSNHG+CEt6RvZEw5V8CH9e3xycOtDe8arWJJwqZ2lC2Ft6SMgPydJFkhczwraLhL3OSJ5NZA5+PRqXR/KYQyDROGi4twMkxHlrWng3aCwnK2lV54xeTbyUApU+cXIU6OD+sqv1WDh7Xphti6hVzC0WyZ96xf9NPQh8pvkQteCCac82g1Ye1n0WowA3xwHxqwQE6aEJV+KKKVKI45aAa9oZKYVTuqa78OmfFkTbGxXKUl4d7t8=">Here is the sketch.</a>

All is very simple, all curves are straight lines, the crossings are found at the correct places.

I suspect that the winding is interpreted incorrectly in this case. For example the dot indicated below has a winding of 2.

![image](https://cloud.githubusercontent.com/assets/2122773/15501377/6f100e02-21ad-11e6-96cf-12ab4cb12a31.png)

It is clearly inside the united path, but the `unite` operator does not define 2 as inside (if I understand this correctly). Here is how the operators are defined in <a href="https://github.com/paperjs/paper.js/blob/develop/src/path/PathItem.Boolean.js#L37">PathItem.Boolean.js</a>

```javascript
    var operators = {
        unite:     { 0: true, 1: true},
        intersect: { 2: true },
        subtract:  { 1: true },
        exclude:   { 1: true }
    };
```

If change the `unite` part to `unite: { 0: true, 1: true, 2: true}`, everything works fine.

@lehni, do you think this could be the solution for this issue?

**Comment by lehni** [comment]
@iconexperience your proposed change breaks about 30 existing unit tests. It's really easy to run these by yourself. Just run this in the paper.js directory:

    gulp test

 Unfortunately I have very little brain capacity nor the time to look into boolean operations again at the moment. The little time I have for paper.js these days I want to focus on finally releasing `v0.10.0`, which is now 3 months delayed. Boolean operations have consumed a disproportionate amount of my time in comparison to everything else that needs taking care of...

**Comment by lehni** [comment]
PS: A point with winding 2 is part of an edge that is fully inside the boolean result, and hence not part of the outlines that will define it.

**Comment by iconexperience** [comment]
@lehni
I fully understand, so I will try to work on these issues myself. I think I understand most of the boolean code now, except for "tracePaths()". Do you know if there is any information on the web that can help me learn more about this part, or should I simply study the code?

**Comment by lehni** [comment]
Sounds good! Didn't I do a write up about tracePaths() a while back?

**Comment by iconexperience** [comment]
Yes, here it is: https://github.com/paperjs/paper.js/issues/874

Thanks!

**Comment by lehni** [comment]
@iconexperience good to have you back BTW! And while I have your attention, I still would like to hear your thoughts on #@590 : )

**Comment by iconexperience** [comment]
Here is another case that probably has the same cause:

```javascript
var c = new CompoundPath();
c.addChild(new Path({segments:[[150, 100], [150, 150], [100, 150], [100, 100]], closed:true}));
c.addChild(new Path({segments:[[200, 150], [200, 100], [220, 100], [220, 150]], closed:true}));
var p =  new Path({segments:[[200, 100], [200, 150], [120, 150], [120, 100]], closed:true})
var res = c.unite(p);
```

<a href="http://sketch.paperjs.org/#S/jVHBasMwDP0V40tTCMEN5JKyU39gsGPag2driYkjGdtZGaX/PnspW8nK2O1JT3pPD104ygl4y19GiGrgJVekc/0uPVPsiSGc2YEmRzPqZxmHYrs/oqqk1ofBWF1k/qt/CdBPgDG0XbdrRMl2QpxKdsPNgsUap5lUKEsBdBv9DNftfwzqO6Fa/JjV9Ro3jw1yPJfisT/kxZ387eZ6jR/cf8RF3kNIBqqa0UQoXHZNrerNWHsgSz6RG2v6IfZkNaAn/QHW0nmzzwqqCtHTCN+jwQCizKT7RTlvsO89AC7L6Y2vHuToyKRAvO1O108=">Here is the sketch</a>

![image](https://cloud.githubusercontent.com/assets/2122773/15576710/6c14ad90-2359-11e6-914f-a01adde3bdeb.png)

**Comment by iconexperience** [comment]
I think the problem here is the way we calculate the winding for a point on a curve if the curve is horizontal at that point. What we currently do is (in pseudo code):

```
if (curve is horizontal at p) {
  find a curve on the same path that crosses a vertical ray through p.x above and/or below p.y
  get the winding of the newsly found point(s) where the vertical ray crosses the curves
}
```

This works in many cases, but note that there is no check if the curves are horizontal at the newly found points. In the examples above we find that the curve is horizontal at that point, search for an alternative point, but fail to notice that the curve at the alternative point is horizontal, too. And that's probably why a wrong winding gets reported.

**Comment by iconexperience** [comment]
Here is another attempt to explain what is going on. In the following examples the two rectangles both have clockwise orientation.

![image](https://cloud.githubusercontent.com/assets/2122773/15635825/28470e9e-25ea-11e6-85b2-c5c895e5c817.png)

If we follow the horizontal ray through point P we find that the winding to the right is 2 and the winding to the left is 0. As a result the `getWinding()` function will return `2`. But I am not sure if this is the correct result. Maybe we should add the following rule to `getWinding()`:

`If windingLeft or windingRight is zero and the other one is not, return 1`

or as code:

`return !windingLeft != !windingRight ? 1 : Math.max(abs(windLeft), abs(windRight));`

This way `getWinding()` will always return `1` for any point on the outside path. I think I tried something similar before, but at that time I made `getWinding()` return `1` for each point on any path, which cause problems with `intersect()` for points on inner paths as in the the following example:

![image](https://cloud.githubusercontent.com/assets/2122773/15635863/9f983d28-25eb-11e6-8476-404575292322.png)

With the idea that I tried previously, the winding for point P would have been ´1´, but with the new idea it is ´2´, which is correct.

**Comment by lehni** [comment]
Interesting! Have you run this change against the unit tests?

**Comment by iconexperience** [comment]
No, but I will. I have the feeling that this will cause issues with `intersect()` operations, because they expect a winding number of 2 if the point is on the result path. Not sure how this all can be solved. Maybe `getWinding()` will have to give different results depending on the type of boolean operation.

**Comment by iconexperience** [comment]
I managed to run the test (could not install cairo, so I ran the tests through the browser using `gulp test:browser`).
Unfortunately as expected there are failing tests (4), all using boolean intersection operations.

As mentioned above, the problem is that if two paths overlap on the outer contour, the overlap should be considered part of the result for `intersect` and `unite`. But `intersect` expects a winding of 2 for points on the result path, whereas `unite` expects `1`. So `getWinding()` has to return different results, depending on the type of boolean operation.

**Comment by iconexperience** [comment]
@lehni Maybe this problem was the reason why you could not <a href="https://github.com/paperjs/paper.js/issues/890#issuecomment-169738849">switch to a single winding check</a> in `propagateWinding()`? Aaargh, the boolean code has become addictive again.

**Comment by iconexperience** [comment]
What I found is that we can change the `unite`operator from `{ 0: true, 1: true }` to `{ 1: true }` without breaking any test cases. This makes sense to me, because a winding of `0` should always be outside the result.

**Comment by lehni** [comment]
@iconexperience did you try adding a separate argument to `getWinding()` and make it behave in the way you outlined above?

**Comment by lehni** [comment]
@iconexperience I think the reason for multiple winding checks had more to do with imprecisions, as far as I can remember.

**Comment by lehni** [comment]
@iconexperience if we start checking the winding operation in `getWinding()`, we should outline exactly how the behaviors should be different. Should we start testing the edge cases for all different operations?

**Comment by lehni** [comment]
@iconexperience I have implemented your suggestion in 648beb33e90821843f1b19143aaaf04f119bcc6d and it fixes these issues while not creating new ones. Are you sure this is the right way of handling it though? We should still make sure it works correctly for all other operators.

**Comment by iconexperience** [comment]
@lehni I am not sure at all. I had my own implementation for testing and it seems to work so far, so it cannot be too wrong. Unfortunately I was not able to find any information on the web about these non-standard cases, so this is all a bit trial-and-error and some sketches with pencil and paper.

**Comment by lehni** [comment]
@iconexperience well, I was wondering if you weren't right about `unite: { 1: true, 2: true}`. It does make sense... But then it creates new issues, which might happen for other reasons. So a part of me thinks that not adding special treatment to `getWinding()` and solving these instead would be the right way to handle this. At the same time, the special treatment does lead to a distinction between a point that's on the border of the union and one that's fully inside...

**Comment by lehni** [comment]
@iconexperience Oh damn. I hate to spoil the fun, but my change in 648beb33e90821843f1b19143aaaf04f119bcc6d unfortunately breaks a *huge* amount of your old path offsetting unit test. I just ran it again it, and I get hundred of glitches. Are you still testing against this also sometimes? : (

**Comment by lehni** [comment]
@iconexperience what's interesting is that the results actually seem fine visually. I just get a ton of "open path" errors. Which may make sense, since it may just mean that more segments are now considered part of the results, and not all of them can be joined up. We may find that there is nothing wrong here and these open paths can just safely be discharged... But more more analysis will be required before we can get to this conclusion. Argh...

**Comment by iconexperience** [comment]
@lehni  I do not think anymore that `unite: { 1: true, 2: true}` makes much sense. In the following example you can see that it would count the blue dot as being part of the result for a unite operation, which does not make much sense:
![image](https://cloud.githubusercontent.com/assets/2122773/15984879/bcb6581a-2fda-11e6-9088-446035314159.png)
Regarding my test suite: No, I have not been running this recently. Currently I am trying to fix the remaining problems in another use case and I am way behind schedule, so I do not have much time to looks at the test suite. But entually it should all work, and all the other changes that were made in the last few days made a lot of sense.
This one is very tricky business, i know this.

**Comment by lehni** [comment]
@iconexperience funny I just came to that same conclusion myself, by testing this and looking at the resulting errors in the paper.js unit tests (they are visualized).

The problem with this change is that it's the biggest breaking change I've experienced in boolean operations so far that produced as many issues in one go.

**Comment by lehni** [comment]
Ok, so here [a rather simple sketch](http://sketch.paperjs.org/#S/dVNNi9xGEP0rYi67Bk2766ur2osPJhDIIRBYgw92DrK2vR5W1iwabTYf+L/7aZbEscGSDq161fVeva7+ZzcPn9ruxe76rq3jx12/G4832//z593tdECoG9Zu/GucWkf6bv5jWLr7Yf1I3ctubo/db1hfXvyqTIndaqFKpqE9W01WmQsFMM1Zxj2VhATNVS3wkkS/V01CRUsQSzbOtduzJBYhNnLh4tn7fVCSYoWxN5jDrNuTJFSNXEwcmIBzTzkQrOImRTkX98lKkgivrCYQwg5pOVkmgVKqRSg2ZSkUXEWrUlCNqr2lXLBAjSpSkNftoctMpAbYxJDuPbRGUUJLlRgbs3cQljVyZkJLEUrel5K0gC2X7HBC5e+LZ1dfveTvvMyewo3A5UEUVnoyShQc4DrrUR81FVXjcJBwLXVrn5O7upXgrDCL9CxGWCs7oROC8f3mr6F+9oBK6IFmtgScSGVLy7mWfi85ieYSruIV+T5xTeTOzuSwScN646TslWoVdtjuIyc0XjUwBiXjUMSgK+Ew2UJEomaHrR0nqVxBhCMsRgE+TqZ4nFWrMbpCUkYgY6pYcTbKWykMTiUB19kZ4ycr383noUyndTnetZ+O03GBqRdLu7m4+ha7HofpMN8C/TBMp3be+3iYb46P6QSo/TyM63lzTnT1H7S2P9dX8y0uwctO7Gt8affHZX2DH9Q8/a/oN/gv89qWUxvXw3H+YdJ1u/3U5vWH+OtlGNuGrsvDk+xtfpZ2ephWRJ96fJgPa7s8D9VmyxOaPhym6V9PtkE7ry+p7/BlfJjrLRuX//3Shrv74wE6di/e/v75Cw==) that now fails but used to work. If you're running it against the [boolean-debug](https://github.com/paperjs/paper.js/tree/boolean-debug) branch, you can use the boolean switches in the code to display some debug info.

**Comment by lehni** [comment]
[And here now properly simplified](http://sketch.paperjs.org/#S/lZNdS8MwFIb/SuiN2+hi06/UDi9EELwQhAlerL2I3XGWZclIMieI/93Tpn5MHGgoNOd9wsl5T5LXQIkNBGUwX4NrnoIwaPSyi59b2NONcKZ9IedEwZ7c9MFoEcc0yRhPGT+LOGdpkYYk6r/fyPSM85zyJCmKnLOC5yglaVLQOGKc8YTnPOL1eFapSj0LQ7aO4X4LkuZZSBiPSD37ALEHKe7Ecg8GJNwTG6q8xfnotVIEh4XVBpSzJVl4oRu4Qdilyfr8GanDLrXndej/jdQWliVxZgeDZJ3Ra7jUUpuSnBhYnhyAeSNkq1YleRTSQqXexgfVxX+vLg6/oq7MD7eHch4N3TmQvTdEU/RF6n+aepA7+JOrSu1btdR7apHDlWicNmgwomz2iRy8uAu1koAgyb50A1tt3D0GmNgi7FP/5NfKgbHQuFaro4vmQwOP8TsjGvhO/XkYsDvpUO6vDd2p1sGoP6TOnaf0sZWy78twcP18hN1l/rLTOOtWn04mp/hmHgyI9Va3WExQLuq3dw==):
```js
view.matrix = new Matrix([22.351741790771484, 0, 0, 22.351741790771484, -9776.73388671876, -3438.201717376707]);

var pt1 = [ 465, 170 ];
var pt2 = [ 440, 160 ];

var path1 = new Path({
    segments: [
        pt1, [ 455, 175 ], pt2
    ],
    closed: true,
    strokeColor: 'red',
    strokeScaling: false
});

var path2 = new Path({
    segments: [
        pt2,
        [ 450, 160 ],
        [ 460, 170 ],
        [ pt1, [0, -5 ] ]
    ],
    closed: true,
    strokeColor: 'blue',
    strokeScaling: false
});

var result = path1.unite(path2);
result.fillColor = new Color(1, 1, 0, 0.25);
```

**Comment by lehni** [comment]
I think it has something to do with winding calculations and `resolveCrossings()`. The blue shape has a self intersection and will be split into two path before the boolean operation is performed.

Here the zoom on the crucial part, with debug information on top. See the blue ids 7.0 and 8.0. Those are the two segments in the location where the blue path was self-intersecting, now they're part of two separate paths (id 7 and 8), and they're the first segment of each path (.0). But one of them has received winding 2, while the other has 1. Why?

<img width="950" alt="screen shot 2016-06-11 at 15 01 13" src="https://cloud.githubusercontent.com/assets/213275/15985184/80277e64-2fe5-11e6-8fb1-2fa6563c8064.png">

**Comment by lehni** [comment]
Oh wait, path 7 is the blue sub-path that goes "down" to the left and overlaps with the red path, hence winding 2 makes sense. path 8 is the blue bump to the right that's merely touching the red path.

The problem is that these segments with winding 1 are lying around and look valid due to the winding value of 1...

**Comment by iconexperience** [comment]
@lehni I am currently busy with another issue, so I do not have much time to look into details. But please consider that an open path is not a problem if it is not required for the final shape. I tried this before and there were many open paths reported, but the results always looked correct. But I am certainly not sure about this.

**Comment by lehni** [comment]
@iconexperience when you have time, check out my recent changes to this fix. It solve this issue, and makes the new issues with your test suite disappear completely. There is currently only one failing case in that suite, which I will look at separately. All in all, great news! I wonder what you think about how I'm handling the winding contributions now in 7241edd98ac135243eab392f63d6ab4f37e07200. I think making this distinction is more elegant, since we don't have to lie about the winding number, and have more information about the nature of the segments available in the boolean code further down.

**Comment by lehni** [comment]
@iconexperience more good news: 693898a5b10e72a56b0eaaf2080bf87b12ac64b7 :)
